ethcore/engines/authority-round/src/randomness.rs

```diff
@@ -0,0 +1,245 @@
+	MissingPublicKey,
+}
+
+impl RandomnessPhase {
+	/// Determine randomness generation state from the contract.
+	///
+	/// Calls various constant contract functions to determine the precise state that needs to be
+	/// handled (that is, the phase and whether or not the current validator still needs to send
+	/// commitments or reveal secrets).
+	pub fn load(
+		contract: &BoundContract,
+		our_address: Address,
+	) -> Result<RandomnessPhase, PhaseError> {
+		// Determine the current round and which phase we are in.
+		let round = contract
+			.call_const(aura_random::functions::current_collect_round::call())
+			.map_err(PhaseError::LoadFailed)?;
+		let is_reveal_phase = contract
+			.call_const(aura_random::functions::is_reveal_phase::call())
+			.map_err(PhaseError::LoadFailed)?;
+		let is_commit_phase = contract
+			.call_const(aura_random::functions::is_commit_phase::call())
+			.map_err(PhaseError::LoadFailed)?;
+
+		// Ensure we are not committing or revealing twice.
+		let committed = contract
+			.call_const(aura_random::functions::is_committed::call(
+				round,
+				our_address,
+			))
+			.map_err(PhaseError::LoadFailed)?;
+		let revealed: bool = contract
+			.call_const(aura_random::functions::sent_reveal::call(
+				round,
+				our_address,
+			))
+			.map_err(PhaseError::LoadFailed)?;
+
+		// With all the information known, we can determine the actual state we are in.
+		if is_reveal_phase && is_commit_phase {
+			return Err(PhaseError::PhaseConflict);
+		}
+
+		if is_commit_phase {
+			if revealed {
+				return Err(PhaseError::RevealedInCommit);
+			}
+
+			if !committed {
+				Ok(RandomnessPhase::BeforeCommit { our_address, round })
+			} else {
+				Ok(RandomnessPhase::Committed)
+			}
+		} else {
+			if !committed {
+				// We apparently entered too late to make a commitment, wait until we get a chance again.
+				return Ok(RandomnessPhase::Waiting);
+			}
+
+			if !revealed {
+				Ok(RandomnessPhase::Reveal { our_address, round })
+			} else {
+				Ok(RandomnessPhase::Waiting)
+			}
+		}
+	}
+
+	/// Advance the randomness state, if necessary.
+	///
+	/// Returns the contract calls necessary to advance the randomness contract's state.
+	///
+	/// **Warning**: The `advance()` function should be called only once per block state; otherwise
+	///              spurious transactions resulting in punishments might be executed.
+	pub fn advance<R: Rng>(
+		self,
+		contract: &BoundContract,
+		rng: &mut R,
+		signer: &dyn EngineSigner,
+	) -> Result<Option<Bytes>, PhaseError> {
+		match self {
+			RandomnessPhase::Waiting | RandomnessPhase::Committed => Ok(None),
+			RandomnessPhase::BeforeCommit { round, our_address } => {
+				// Check whether a secret has already been committed in this round.
+				let committed_hash: Hash = contract
+					.call_const(aura_random::functions::get_commit::call(round, our_address))
+					.map_err(PhaseError::LoadFailed)?;
+				if !committed_hash.is_zero() {
```

Comment:
Yes, I'm happy to remove it.
This code was written very defensively because it's supposed to work with randomness contract implementations that are not part of this repository. It would be a bug, but a contract could, of course, trigger this error.

Reply:
The contract takes into account this (see the implementation of `commitHash` function in https://github.com/paritytech/parity-ethereum/pull/10946#issuecomment-548404697).

Reply:
That check was removed.